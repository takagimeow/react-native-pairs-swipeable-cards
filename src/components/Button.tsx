import React from 'react';
import { View, TouchableOpacity, StyleSheet } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';

export const Button = ({
  name,
  color = '#c4c4c4',
  onPress,
}: {
  name: 'reply' | 'heart' | 'handshake-o';
  color?: string;
  onPress?: () => void;
}) => {
  return (
    <TouchableOpacity style={styles.container} onPress={onPress}>
      <View style={styles.iconContainer}>
        <FontAwesome name={name} size={28} color={color} />
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#ffffff',
    elevation: 5,
    borderRadius: 50,
    shadowColor: 'rgba(196, 196, 196, 0.25)',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 2.5,
    shadowRadius: 7.4,
    paddingHorizontal: 8,
    paddingVertical: 8,
  },
  iconContainer: {
    width: 48,
    height: 48,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
