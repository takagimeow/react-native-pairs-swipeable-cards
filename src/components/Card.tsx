import React from 'react';
import { Animated, View, StyleSheet, Dimensions, PanResponderInstance, Text } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';

export const CARD_WIDTH = Dimensions.get('screen').width * 0.915;
export const CARD_HEIGHT = Dimensions.get('screen').height * 0.695;
export const BORDER_RADIUS = 12;

export interface ICard {
  children: JSX.Element;
  panResponder: PanResponderInstance;
  pan: Animated.ValueXY;
  opacity: Animated.Value;
  overlayOpacity: Animated.AnimatedInterpolation;
  overlayColor: Animated.AnimatedInterpolation;
  cardState: string;
  rotate: string | Animated.Value | Animated.AnimatedInterpolation;
  scale: number | Animated.Value | Animated.AnimatedInterpolation;
}

export const Card = ({
  children,
  panResponder,
  pan,
  opacity,
  overlayOpacity,
  overlayColor,
  cardState,
  rotate,
  scale,
}: ICard) => {
  return (
    <Animated.View
      {...panResponder.panHandlers}
      style={[
        pan.getLayout(),
        styles.box,
        {
          opacity,
          transform: [
            {
              rotate,
            },
            {
              scale: scale,
            },
          ],
        },
        {
          position: 'absolute',
        },
      ]}
    >
      <Animated.View
        style={[
          {
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            backgroundColor: overlayColor,
            zIndex: 1,
            opacity: overlayOpacity,
            paddingHorizontal: 8,
            paddingVertical: 8,
            borderRadius: BORDER_RADIUS - 1,
          },
        ]}
      >
        <View
          style={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: cardState === 'good' ? 'flex-start' : 'flex-end',
            alignItems: 'center',
          }}
        >
          <View
            style={{
              marginRight: 8,
            }}
          >
            <FontAwesome
              name={cardState === 'good' ? 'thumbs-up' : 'reply'}
              size={40}
              color={'#ffffff'}
            />
          </View>
          <Text style={styles.text}>{cardState === 'good' ? 'いいね' : 'スキップ'}</Text>
        </View>
      </Animated.View>
      {children}
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  box: {
    backgroundColor: '#f0f0f0',
    width: CARD_WIDTH,
    height: CARD_HEIGHT,
    borderRadius: BORDER_RADIUS,
    borderWidth: 1,
  },
  text: {
    fontSize: 48,
    fontWeight: 'bold',
    color: '#ffffff',
  },
});
