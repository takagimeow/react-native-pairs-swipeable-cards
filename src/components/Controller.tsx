import React from 'react';
import { StyleSheet, View, Dimensions } from 'react-native';

import { Button } from './Button';

export interface IController {
  onSkip?: () => void;
  onLikeWithMessage?: () => void;
  onLike?: () => void;
}

export const Controller = ({ onSkip, onLikeWithMessage, onLike }: IController) => {
  return (
    <View style={styles.container}>
      <Button name="reply" onPress={onSkip} />
      <Button name="handshake-o" color="#fc5766" onPress={onLikeWithMessage} />
      <Button name="heart" color="#fc5766" onPress={onLike} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
    width: Dimensions.get('window').width,
    paddingBottom: 32,
    paddingHorizontal: 64,
  },
});
