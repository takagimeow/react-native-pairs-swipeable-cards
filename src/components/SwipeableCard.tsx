import React, { useEffect, useRef, useState } from 'react';
import { Animated, PanResponder, Dimensions } from 'react-native';
import { Card } from './Card';

export function isDropArea(moveX: number) {
  const screenWidth = Dimensions.get('screen').width;
  if (moveX < -(screenWidth * 0.4) || moveX > screenWidth * 0.7) {
    return true;
  }
  return false;
}

export type Props = {
  id: number;
  children: React.ReactElement;
  skipped: number;
  liked: number;
  current?: boolean;
  next?: boolean;
  scale?: Animated.Value | Animated.AnimatedInterpolation;
  onSkip?: (id: number) => void;
  onLike?: (id: number) => void;
  onPanResponderMove: (id: number, dx: number, dy: number) => void;
};

export function SwipeableCard({
  id,
  children,
  skipped,
  liked,
  current,
  next,
  scale,
  onSkip,
  onLike,
  onPanResponderMove,
}: Props) {
  const pan = useRef(new Animated.ValueXY()).current;
  const opacityRef = useRef(new Animated.Value(1));
  const [visibility, setVisibility] = useState(true);
  const [cardState, setCardState] = useState('skip');
  // これで、左下から右上に移動するとき、本来ならpan.xは正の値になるが、
  // pan.x（プラス）とpan.y（マイナス）を掛け合わせているので、-20degが設定される
  // 加えて、右下から左上に移動するとき、本来ならpan.xは負の値になるが、
  // pan.x（マイナス）とpan.y（マイナス）を掛け合わせているので、20degが設定される
  const rotateValue = Animated.multiply(pan.x, pan.y).interpolate({
    inputRange: [-18000, 0, 18000],
    outputRange: ['-20deg', '0deg', '20deg'],
    extrapolate: 'clamp',
  });
  const overlayOpacity = pan.x.interpolate({
    inputRange: [-360, 0, 360],
    outputRange: [1, 0, 1],
    extrapolate: 'clamp',
  });
  const overlayColor = pan.x.interpolate({
    inputRange: [-360, 0, 360],
    outputRange: ['#f0f0f0', 'transparent', '#fc5766'],
    extrapolate: 'clamp',
  });

  const horizontalOffset = useRef(0);

  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderMove: Animated.event(
        [
          null,
          {
            dx: pan.x, // x,y are Animated.Value
            dy: pan.y,
          },
        ],
        {
          listener: _ => {
            if (onPanResponderMove) {
              onPanResponderMove(id, (pan.x as any)._value, (pan.y as any)._value);
            }
            if ((pan.x as any)._value > 0) {
              setCardState('good');
            } else {
              setCardState('skip');
            }
          },
          useNativeDriver: false,
        },
      ),
      onPanResponderRelease: (_, gesture) => {
        if (isDropArea(gesture.dx)) {
          horizontalOffset.current = gesture.dx;
          if (horizontalOffset.current < 0 && onSkip) {
            onSkip(id);
          }
          if (horizontalOffset.current > 0 && onLike) {
            onLike(id);
          }
          Animated.parallel([
            Animated.timing(opacityRef.current, {
              toValue: 0,
              duration: 450,
              useNativeDriver: false,
            }),
            Animated.spring(pan, {
              toValue: {
                x: gesture.dx * 1.25,
                y: 50,
              },
              useNativeDriver: false,
            }),
          ]).start(({}) => {
            setVisibility(false);
          });
        } else {
          Animated.spring(
            pan, // Auto-multiplexed
            { toValue: { x: 0, y: 0 }, useNativeDriver: false }, // Back to zero
          ).start();
        }
      },
    }),
  ).current;

  useEffect(() => {
    const token = pan.x.addListener(({ value }) =>
      onPanResponderMove(id, value, (pan.y as any)._value),
    );
    return () => {
      pan.x.removeListener(token);
    };
  }, []);

  useEffect(() => {
    if (skipped === id) {
      setCardState('skip');
      skip();
      return;
    }
    if (liked === id) {
      setCardState('good');
      like();
      return;
    }
  }, [skipped, liked]);

  const skip = () => {
    Animated.parallel([
      Animated.timing(opacityRef.current, {
        toValue: 0,
        duration: 500,
        useNativeDriver: false,
      }),
      Animated.spring(pan, {
        toValue: {
          x: (Dimensions.get('window').width / 2) * -1,
          y: 50,
        },
        useNativeDriver: false,
      }),
    ]).start(({}) => {
      setVisibility(false);
    });
  };

  const like = () => {
    Animated.parallel([
      Animated.timing(opacityRef.current, {
        toValue: 0,
        duration: 500,
        useNativeDriver: false,
      }),
      Animated.spring(pan, {
        toValue: {
          x: Dimensions.get('window').width / 2,
          y: 50,
        },
        useNativeDriver: false,
      }),
    ]).start(({}) => {
      setVisibility(false);
    });
  };

  if (!visibility) {
    return null;
  }
  return (
    <Card
      panResponder={panResponder}
      pan={pan}
      opacity={opacityRef.current}
      overlayOpacity={overlayOpacity}
      overlayColor={overlayColor}
      cardState={cardState}
      rotate={rotateValue}
      scale={current ? new Animated.Value(1) : next && scale ? scale : 0.925}
    >
      {children}
    </Card>
  );
}
