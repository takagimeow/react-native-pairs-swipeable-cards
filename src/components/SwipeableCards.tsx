import React, { useCallback, useEffect, useRef, useState } from 'react';
import { Animated, StyleSheet, View } from 'react-native';
import { SwipeableCard } from './SwipeableCard';
import { CARD_HEIGHT, CARD_WIDTH } from './Card';

import { Controller } from './Controller';

export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms, 'sleep'));
}

export interface Item {
  id: string;
}

export type Props<T extends Item> = {
  data: T[];
  renderItem: (item: T, index: number) => React.ReactElement;
  onSkip?: (item: T) => void;
  onLike?: (item: T) => void;
  onLikeWithMessage?: (item: T) => void;
};

export function SwipeableCards<T extends Item>({
  data,
  renderItem,
  onLike,
  onLikeWithMessage,
  onSkip,
}: Props<T>) {
  const [reversedData, setReversedData] = useState<T[]>([]);
  const [currentCardIndex, setCurrentCardIndex] = useState(data.length - 1);
  const currnetCardDx = useRef(new Animated.Value(0));
  const currentCardDy = useRef(new Animated.Value(0));

  const [skipped, setSkipped] = useState(-1);
  const [liked, setLiked] = useState(-1);
  const timerIdRef = useRef<NodeJS.Timeout>();

  const [scale] = useState(
    currnetCardDx.current.interpolate({
      inputRange: [-180, 0, 180],
      outputRange: [1, 0.925, 1],
      extrapolate: 'clamp',
    }),
  );

  useEffect(() => {
    data.reverse();
    setReversedData(data);
  }, []);

  const handlePanResponderMove = useCallback((_, dx, dy) => {
    currnetCardDx.current.setValue(dx);
    currentCardDy.current.setValue(dy);
  }, []);

  const next = useCallback(
    (cardId: number, from?: string) => {
      if (timerIdRef.current) {
        return;
      }
      if (from && from === 'SKIP') {
        setSkipped(cardId);
        timerIdRef.current = setTimeout(() => {
          setCurrentCardIndex(cardId - 1);
          timerIdRef.current = undefined;
        }, 500);
        onSkip && onSkip(reversedData[cardId]);
        return;
      }
      if (from && from === 'LIKE') {
        setLiked(cardId);
        timerIdRef.current = setTimeout(() => {
          setCurrentCardIndex(cardId - 1);
          timerIdRef.current = undefined;
        }, 500);
        onLike && onLike(reversedData[cardId]);
        return;
      }
      if (from && from === 'LIKE_WITH_MESSAGE') {
        setLiked(cardId);
        timerIdRef.current = setTimeout(() => {
          setCurrentCardIndex(cardId - 1);
          timerIdRef.current = undefined;
        }, 500);
        onLikeWithMessage && onLikeWithMessage(reversedData[cardId]);
        return;
      }
      if (from && from === 'SKIP_MANUALLY' && onSkip) {
        onSkip(reversedData[cardId]);
      }
      if (from && from === 'LIKE_MANUALLY' && onLike) {
        onLike(reversedData[cardId]);
      }
      setCurrentCardIndex(cardId - 1);
    },
    [currentCardIndex, setSkipped, setLiked],
  );

  return (
    <View style={styles.container}>
      <View style={styles.cardContainer}>
        <View style={styles.box}>
          {reversedData.map((item, index) => {
            return (
              <SwipeableCard
                key={`key_${index}`}
                id={index}
                skipped={skipped}
                liked={liked}
                current={currentCardIndex === index}
                next={currentCardIndex - 1 === index}
                scale={currentCardIndex - 1 === index ? scale : undefined}
                onPanResponderMove={handlePanResponderMove}
                onSkip={id => next(id, 'SKIP_MANUALLY')}
                onLike={id => next(id, 'LIKE_MANUALLY')}
              >
                {renderItem(item, index)}
              </SwipeableCard>
            );
          })}
        </View>
      </View>
      <Controller
        onSkip={() => {
          next(currentCardIndex, 'SKIP');
        }}
        onLike={() => {
          next(currentCardIndex, 'LIKE');
        }}
        onLikeWithMessage={() => {
          next(currentCardIndex, 'LIKE_WITH_MESSAGE');
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'flex-end',
    alignItems: 'center',
    height: '100%',
  },
  cardContainer: {
    position: 'relative',
    width: 'auto',
    height: 'auto',
    marginBottom: 48,
  },
  box: {
    position: 'relative',
    width: CARD_WIDTH,
    height: CARD_HEIGHT,
  },
});
