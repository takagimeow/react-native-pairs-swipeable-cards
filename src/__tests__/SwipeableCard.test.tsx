/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import { View } from 'react-native';
import { render } from '@testing-library/react-native';

import { SwipeableCard } from '../components/SwipeableCard';

jest.mock('@expo/vector-icons', () => {
  return {
    FontAwesome: () => <View></View> as any,
  };
});
describe('SwipeableCard', () => {
  it('Snapshot Test', () => {
    const { toJSON } = render(
      <SwipeableCard
        key={`key_${0}`}
        id={0}
        skipped={-1}
        liked={-1}
        current={true}
        next={false}
        scale={undefined}
        onPanResponderMove={(id, dx, dy) => {
          console.log('id: ', id);
          console.log('dx: ', dx);
          console.log('dy: ', dy);
        }}
        onSkip={id => console.log('skip: ', id)}
        onLike={id => console.log('like: ', id)}
      >
        <View></View>
      </SwipeableCard>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
