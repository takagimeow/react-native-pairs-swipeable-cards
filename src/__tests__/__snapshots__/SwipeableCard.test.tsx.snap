// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeableCard Snapshot Test 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "backgroundColor": "#f0f0f0",
      "borderRadius": 12,
      "borderWidth": 1,
      "height": 927.1299999999999,
      "left": 0,
      "opacity": 1,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "rotate": "0deg",
        },
        Object {
          "scale": 1,
        },
      ],
      "width": 686.25,
    }
  }
>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0)",
        "borderRadius": 11,
        "height": "100%",
        "left": 0,
        "opacity": 0,
        "paddingHorizontal": 8,
        "paddingVertical": 8,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        style={
          Object {
            "marginRight": 8,
          }
        }
      >
        <View />
      </View>
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 48,
            "fontWeight": "bold",
          }
        }
      >
        スキップ
      </Text>
    </View>
  </View>
  <View />
</View>
`;
